### PR TITLE
Refactor: split parsing logic from Execer

### PR DESCRIPTION
**Pull request by agoose77**
Currently the Execer contains parsing logic that handles the contextual resolution of Xonsh syntax. This PR separates this logic into a new `Parser` class that serves as the main parser interface.

The only backwards incompatible change is that the `xonsh.parser.Parser` object is now a different class with a different API. This can be changed if necessary.

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### CODE REVIEW

File: xonsh/lexer.py [hunk 1]
```diff
@@ -408,10 +407,39 @@ def _new_token(type, value, pos):
     return o
 
 
+def _get_token_types():
+    kwlist = kwmod.kwlist[:]
+    if PYTHON_VERSION_INFO >= (3, 9, 0) and PYTHON_VERSION_INFO < (3, 10):
+        kwlist.remove("__peg_parser__")
```
Comment (jnoortheen) [inline xonsh/lexer.py:413]: This can be simply `version == 3.9` ?

Comment (agoose77) [reply xonsh/lexer.py:413]: Yes, simply `sys.version_info[:2] == (3, 9). Note that I didn't write this code - it was just moved. I don't know why we have our own variable for this.

File: xonsh/parsers/base.py [hunk 2]
```diff
@@ -238,20 +237,6 @@ def raise_parse_error(
...
-        super().__init__(*args, **kwargs)
-        self.daemon = True
-        self.parser = parser
-        self.yacc_kwargs = yacc_kwargs
-        self.start()
-
-    def run(self):
-        self.parser.parser = yacc.yacc(**self.yacc_kwargs)
-
-
```
Comment (jnoortheen) [inline xonsh/parsers/base.py:254]: Does this mean, the parser is loaded eagerly? I have all my config in pure Python and parser is not needed for the startup making it faster with threaded version

**Comment by jnoortheen** [comment]
@agoose77 are you still working on this?

**Comment by agoose77** [comment]
> @agoose77 are you still working on this?

Hi! I'm not _right_ now as I'm not really able to spare many development cycles at the moment. I  hope to pick it up in the future. However if you need / want to work on any of my PRs, you're more than welcome! Thanks.

**Comment by jnoortheen** [comment]
> I hope to pick it up in the future. 

There are file conflicts. It needs to be rebased on latest changes. Also the way it loads the parser need to be threaded too. I am closing it for now. Please feel free to re-open or create new one when it is ready.

**Pull request closed without merging**